crypto/rand/rand_lib.c

```diff
@@ -211,25 +213,331 @@ void rand_cleanup_int(void)
+ * the input has 'entropy_per_byte' entropy bits per byte.
+ * In case
+ */
+
+size_t RAND_POOL_bytes_needed(RAND_POOL *pool, unsigned int entropy_per_byte)
+{
+    size_t bytes_needed;
+
+    size_t entropy_needed = RAND_POOL_entropy_needed(pool);
+
+    if (entropy_per_byte < 1 || entropy_per_byte > 8) {
+        RANDerr(RAND_F_RAND_POOL_BYTES_NEEDED, RAND_R_ARGUMENT_OUT_OF_RANGE);
+        return 0;
+    }
+
+    bytes_needed = ENTROPY_TO_BYTES(entropy_needed, entropy_per_byte);
+
+    if (bytes_needed > pool->max_len - pool->len) {
+        /* not enough space left */
+        RANDerr(RAND_F_RAND_POOL_BYTES_NEEDED, RAND_R_RANDOM_POOL_OVERFLOW);
+        return 0;
+    }
+    else if (pool->len < pool->min_len &&
+             bytes_needed < pool->min_len - pool->len)
+        /* to meet the min_len requirement */
+        bytes_needed = pool->min_len - pool->len;
+
+    return bytes_needed;
+}
+
+/*
+ * Returns the remaining number of bytes available
+ *
+ * If an error occurs, -1 is returned.
+ */
+
+size_t RAND_POOL_bytes_remaining(RAND_POOL *pool)
+{
+    return pool->max_len - pool->len;
+}
+
+/*
+ * Add the contents of the |buffer| to the random pool.
+ * Return available amount of entropy after this operation.
+ * (see RAND_POOL_entropy_available(pool))
+ */
+size_t RAND_POOL_add(RAND_POOL *pool,
+                     const unsigned char *buffer, size_t len, size_t entropy)
+{
+    if (len > pool->max_len - pool->len) {
+        RANDerr(RAND_F_RAND_POOL_ADD, RAND_R_ENTROPY_INPUT_TOO_LONG);
+        return 0;
+    }
+
+    if (len > 0) {
+        memcpy(pool->buffer + pool->len, buffer, len);
+        pool->len += len;
+        pool->entropy += entropy;
+    }
+
+    return RAND_POOL_entropy_available(pool);
+}
+
+/*
+ * Add the contents of the |buffer| to the random pool in-place.
+ * Return available amount of entropy after this operation.
+ * (see RAND_POOL_entropy_available(pool))
+ *
+ * RAND_POOL_add_begin() reserves the next |num| bytes for adding
+ * randomness in-place and returns a pointer to the buffer. It
+ * is allowed to copy up to |num| bytes into the buffer.
+ * If |num| == 0 this is considered a no-op and a NULL pointer
+ * is returned without producing an error message.
+ *
+ * RAND_POOL_add_end() is called after updating the buffer. The
+ * |num| argument specifies the number of bytes updated. It is
+ * allowed to update less bytes than originally reserved.
+ */
+unsigned char * RAND_POOL_add_begin(RAND_POOL *pool, size_t len)
+{
+    if (pool->reserved_len != 0) {
+        RANDerr(RAND_F_RAND_POOL_ADD_BEGIN, RAND_R_NESTED_GROUP);
+        return NULL;
+    }
+
+    if (len == 0)
+        return NULL;
+
+    if (pool->len + len > pool->max_len) {
+        RANDerr(RAND_F_RAND_POOL_ADD_BEGIN, RAND_R_RANDOM_POOL_OVERFLOW);
+        return NULL;
+    }
+
+    pool->reserved_len = len;
+
+    return pool->buffer + pool->len;
+}
+
+size_t RAND_POOL_add_end(RAND_POOL *pool, size_t len, size_t entropy)
```

Comment:
There is a lengthy explanation in the `rand_drbg_st` struct, saying that every variable that has the 'len' suffix is a length in bytes, and entropy is counted in bits. 

```
    /*
     * The following parameters are setup by the per-type "init" function.
     *
     * Currently the only type is CTR_DRBG, its init function is ctr_init().
     *
     * The parameters are closely related to the ones described in
     * section '10.2.1 CTR_DRBG' of [NIST SP 800-90Ar1], with one
     * crucial difference: In the NIST standard, all counts are given
     * in bits, whereas in OpenSSL entropy counts are given in bits
     * and buffer lengths are given in bytes.
     *
     * Since this difference has lead to some confusion in the past,
     * (see [GitHub Issue #2443], formerly [rt.openssl.org #4055])
     * the 'len' suffix has been added to all buffer sizes for
     * clarification.
     */
```

Reply:
Yup, that helped me remember the difference.  I don't know if we want to rely on all readers of the code having (first?) read the comment at that structure, though.

Reply:
I reformulated the comments before `RAND_POOL_add*()`, trying to be more precise about bits & bytes.